### ISSUE TITLE
Venv mode is broken with concurrency when running with Black on macOS

### ISSUE THREAD

**Issue by Eric-Arellano** [pantsbuild/pex#1235]
This fails: `pex --venv=append -m black:patched_main --interpreter-constraint='CPython>=3.6' black==20.8b1 -o black.pex`, then `./black.pex **/*.py` in a repo with ~ >10 Python files, e.g. running against `src/python/pants/**/*.py` in pantsbuild/pants.

<details><summary>Traceback</summary>

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/pex", line 104, in <module>
    func()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 6607, in patched_main
    main()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 782, in main
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    rv = self.invoke(ctx)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 610, in invoke
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    return callback(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    exitcode = _main(fd, parent_sentinel)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    return f(get_current_context(), *args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 556, in main
    prepare(preparation_data)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 87, in _run_code
    reformat_many(
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 705, in reformat_many
    exec(code, run_globals)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/pex", line 104, in <module>
    func()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 6607, in patched_main
    loop.run_until_complete(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 755, in schedule_formatting
    tasks = {
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 757, in <dictcomp>
    main()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 782, in main
    loop.run_in_executor(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    rv = self.invoke(ctx)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    executor.submit(func, *args), loop=self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 697, in submit
    self._adjust_process_count()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 675, in _adjust_process_count
    p.start()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    self._launch(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    return callback(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    return f(get_current_context(), *args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 556, in main
    reformat_many(
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 705, in reformat_many
    loop.run_until_complete(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 755, in schedule_formatting
    tasks = {
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 757, in <dictcomp>
    loop.run_in_executor(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 697, in submit
    self._adjust_process_count()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 675, in _adjust_process_count
    p.start()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/pex", line 104, in <module>
    func()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 6607, in patched_main
    main()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 556, in main
    reformat_many(
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 705, in reformat_many
    loop.run_until_complete(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 755, in schedule_formatting
    tasks = {
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 757, in <dictcomp>
    loop.run_in_executor(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 697, in submit
    self._adjust_process_count()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 675, in _adjust_process_count
    p.start()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(Traceback (most recent call last):
process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
  File "<string>", line 1, in <module>
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/pex", line 104, in <module>
    func()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 6607, in patched_main
    main()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 556, in main
    reformat_many(
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 705, in reformat_many
    loop.run_until_complete(
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    return future.result()
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 755, in schedule_formatting
    main_content = runpy.run_path(main_path,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/pex", line 104, in <module>
    tasks = {
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 757, in <dictcomp>
    func()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 6607, in patched_main
    loop.run_in_executor(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 697, in submit
    main()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    self._adjust_process_count()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 675, in _adjust_process_count
    return self.main(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    p.start()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/process.py", line 121, in start
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    self._popen = self._Popen(self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    return callback(*args, **kwargs)
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    super().__init__(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    return f(get_current_context(), *args, **kwargs)
    self._launch(process_obj)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 42, in _launch
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 556, in main
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
    reformat_many(
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 705, in reformat_many
    loop.run_until_complete(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 755, in schedule_formatting
error: cannot format /Users/eric/code/pants/src/python/pants/backend/awslambda/python/rules_test.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/codegen/protobuf/python/rules.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/python/dependency_inference/module_mapper.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/codegen/protobuf/__init__.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/project_info/filter_targets_test.py: A process in the process pool was terminated abruptly while the future was     tasks = {
running or pending.
  File "/Users/eric/.pex/venvs/068d8301e3a009704ad6ad1c19b1960662ea2bea/837c093c06bd686fc4cd8a7c11bdb7b9cb3598d4/lib/python3.9/site-packages/black/__init__.py", line 757, in <dictcomp>
    loop.run_in_executor(
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/concurrent/futures/process.py", line 697, in submit
    self._adjust_process_count()
error: cannot format /Users/eric/code/pants/src/python/pants/backend/python/dependency_inference/module_mapper_test.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/__main__.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/codegen/protobuf/python/rules_integration_test.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/codegen/protobuf/protobuf_dependency_inference.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/project_info/list_roots.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/codegen/protobuf/target_types.py: A process in the process pool was terminated abruptly while the future was running or pending.
error: cannot format /Users/eric/code/pants/src/python/pants/backend/__init__.py: A process in the process pool was terminated abruptly while the future was running or pending.
... <above error repeated for every file>
Oh no! 💥 💔 💥
32 files left unchanged, 362 files failed to reformat.
/Users/eric/.pyenv/versions/3.9.1/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 20 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

</details>

It also fails with `--venv=prepend`, but it does not fail with `--unzip`. It does not fail on Ubuntu.

**Comment by stuhood** [comment]
Yea, I repro. Attempting to format the files in the `pex` repro is sufficient.

**Comment by stuhood** [comment]
> Yea, I repro on macOS. Attempting to format the files in the `pex` repro is sufficient.

Oh. But only with `3.9.*`. It's fine with Python `3.7.7`.

**Comment by Eric-Arellano** [comment]
> Oh. But only with 3.9.*. It's fine with Python 3.7.7. Which might suggest an API change in that concurrency API / black rather than anything PEX did.

Interesting. Confirmed that it still works with Python 3.9 on Ubuntu.

So something about macOS + 3.9 + venv mode rather than unzip / zip app.

**Comment by jsirois** [comment]
OK - The description here turned out to be pretty far off. The issue appears to be the lack of a `if __name__ == '__main__'` gaurd in the `pex` venv script. That doesn't play well with multiprocessing on macOS for Python 3.9.

**Comment by jsirois** [comment]
I'm going to use #1236 to track the narrow problem / for the CHANGES.rst and close this when that's released.

**Comment by jsirois** [comment]
This is now fixed in Pants with its upgrade to Pex 2.1.31 in https://github.com/pantsbuild/pants/commit/0e54e27eaae4ad338c25fb4d7c25dc2afa4a4f17.
